src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageSource.java

```diff
@@ -0,0 +1,660 @@
+	}
+
+	@Override
+	public String getComponentType() {
+		return "kafka:message-source";
+	}
+
+	@Override
+	protected synchronized Object doReceive() {
+		if (this.consumer == null) {
+			createConsumer();
+		}
+		ConsumerRecord<K, V> record;
+		TopicPartition topicPartition;
+		synchronized (this.consumerMonitor) {
+			Set<TopicPartition> paused = this.consumer.paused();
+			if (paused.size() > 0) {
+				this.consumer.resume(paused);
+			}
+			ConsumerRecords<K, V> records = this.consumer.poll(this.pollTimeout);
+			this.consumer.pause(this.partitions);
+			if (records == null || records.count() == 0) {
+				return null;
+			}
+			record = records.iterator().next();
+			topicPartition = new TopicPartition(record.topic(), record.partition());
+			if (records.count() > 1) {
+				this.consumer.seek(topicPartition, record.offset() + 1);
+			}
+		}
+		Producer<K, V> producer = null;
+		AtomicBoolean myTransaction = new AtomicBoolean();
+		AtomicReference<Producer<K, V>> producerRef = new AtomicReference<>();
+		KafkaAckInfo<K, V> ackInfo = new KafkaAckInfo<K, V>(this.consumerMonitor, this.groupId, this.consumer, record,
+				topicPartition, producerRef, this.inflightRecords, myTransaction);
+		if (this.producerFactory != null) {
+			@SuppressWarnings("unchecked")
+			KafkaResourceHolder<K, V> holder = (KafkaResourceHolder<K, V>) TransactionSynchronizationManager
+					.getResource(this.producerFactory);
+			if (holder == null) {
+				holder = createTransaction(ackInfo);
+				myTransaction.set(true);
+			}
+			producer = holder.getProducer();
+			producerRef.set(producer);
+		}
+		AcknowledgmentCallback ackCallback = this.ackCallbackFactory.createCallback(ackInfo);
+		this.inflightRecords.computeIfAbsent(topicPartition, tp -> new TreeSet<>()).add(ackInfo);
+		@SuppressWarnings("unchecked")
+		Message<Object> message = (Message<Object>) this.messageConverter.toMessage(record,
+				ackCallback instanceof Acknowledgment ? (Acknowledgment) ackCallback : null, this.consumer,
+				this.payloadType);
+		if (message.getHeaders() instanceof KafkaMessageHeaders) {
+			Map<String, Object> rawHeaders = ((KafkaMessageHeaders) message.getHeaders()).getRawHeaders();
+			rawHeaders.put(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK, ackCallback);
+			if (this.producerFactory != null) {
+				rawHeaders.put(PRODUCER_FACTORY, ackInfo);
+			}
+			return message;
+		}
+		else {
+			AbstractIntegrationMessageBuilder<Object> builder = getMessageBuilderFactory().fromMessage(message)
+					.setHeader(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK, ackCallback);
+			if (this.producerFactory != null) {
+				builder.setHeader(PRODUCER_FACTORY, ackInfo);
+			}
+			return builder;
+		}
+	}
+
+	private KafkaResourceHolder<K, V> createTransaction(KafkaAckInfo<K, V> ackInfo) {
+		Producer<K, V> producer = this.producerFactory.createProducer();
+		producer.beginTransaction();
+		KafkaResourceHolder<K, V> resourceHolder = new KafkaResourceHolder<K, V>(producer);
```

Comment:
Why `ProducerFactoryUtils.getTransactionalResourceHolder()` doesn't work here for us?

Reply:
Because we need a different binding for each transaction - there might be multiple transactions in flight at the same time.

Reply:
Not sure how it is related to the thread-bound transaction model. We start a TX before `receive()`, process a message (if any) and return the call to the after `receive()` for TX commit or rallback. 
How one may have several resources bounded to the same thread at the same time when we have only one object (message) during `receive()` call? Well, you may store several resources to the thread local from different calls, indeed, but after you finish the first one, the TX will be committed or rolled back and the resource will be unbound from thread-local.

What am I missing?

Sorry for my ignorance but since I'm in charge for review this I would like to understand how things works and are going to work when this code goes to the `master`. 

Thank you!

Reply:
I was trying to solve this use case...

- main thread polls the source several times
- after each poll, hands off the work to other threads
- when the tasks complete, they signal back to the main thread which acks the messages

Hence, the possibility of multiple in-flight transactions.

I suppose it would be simpler to not support that use case; the user can use multiple threads to poll the source instead.

Reply:
Ah! I see. But still doesn't matter that we should store each `KafkaAckInfo` into thread local.
Or we have a global TX around all the `receive()` calls there or TX per `receive()`. In any case we should defer commit or rollback to the end of TX independently of the order for messages processing. Well, I even would say it doesn't make sense to populate the acknowledge in case of transactions otherwise we really break a purpose of transactions when we may ack/nack offsets individually without impact for the rest of records in the same TX. 

Just my vision...